streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java

```diff
@@ -1185,7 +1262,26 @@ public void process(final ProcessorSupplier<? super K, ? super V> processorSuppl
 
             // do not have serde for joined result;
             // also for key serde we do not inherit from either since we cannot tell if these two serdes are different
-            return new KStreamImpl<>(joinMergeName, joined.keySerde(), null, allSourceNodes, false, joinGraphNode, builder);
+            return new KStreamImpl<>(joinMergeName, streamJoinedInternal.keySerde(), null, allSourceNodes, false, joinGraphNode, builder);
+        }
+
+        private void assertWindowSettings(final WindowBytesStoreSupplier supplier, final JoinWindows joinWindows) {
+            final long joinGrace = joinWindows.gracePeriodMs() < 0 ? 0 : joinWindows.gracePeriodMs();
+            final boolean allMatch = supplier.retentionPeriod() == (joinWindows.size() + joinGrace) &&
+                supplier.windowSize() == joinWindows.size() && supplier.retainDuplicates();
```

Comment:
Should we validate `supplier.retainDuplicates()` separately, so that the error message can specifically say that you have to assert `retainDuplicates`?

Reply:
ack